src/go/k8s/pkg/resources/secret.go

```diff
@@ -0,0 +1,247 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+// Key returns namespace/name object that is used to identify object.
+// For reference please visit types.NamespacedName docs in k8s.io/apimachinery
+func (r *SecretResource) Key() types.NamespacedName {
+	return SecretKey(r.pandaCluster)
+}
+
+// SecretKey provides config map name that derived from redpanda.vectorized.io CR
+func SecretKey(pandaCluster *redpandav1alpha1.Cluster) types.NamespacedName {
+	return types.NamespacedName{Name: resourceNameTrim(pandaCluster.Name, baseSuffix), Namespace: pandaCluster.Namespace}
+}
+
+func secretKind() string {
+	var cfg corev1.Secret
+	return cfg.Kind
+}
+
+// getPostStartScript creates a script that removes maintenance mode after startup.
+func (r *SecretResource) getPostStartScript() string {
+	// TODO replace scripts with proper RPK calls
+	curlNodeIDCommand := r.composeCURLGetNodeIDCommand("--silent --fail")
+	curlCommand := r.composeCURLMaintenanceCommand(`-X DELETE --silent -o /dev/null -w "%{http_code}"`, nil)
+	// HTTP code 400 is returned by v22 nodes during an upgrade from v21 until the new version reaches quorum and the maintenance mode feature is enabled
+	// TODO(joe): The command builder shouldn't add auth/tls flags, they should be added to an environment variable
+	// in the StatefulSet otherwise a race condition could occur where the mounted script could be updated before
+	// the new setting has been configured on the pod's redpanda node.
+	return fmt.Sprintf(`#!/usr/bin/env bash
+set -e
+
+until NODE_ID=$(%s | grep -o '\"node_id\":[^,}]*' | grep -o '[^: ]*$'); do
+	sleep 0.5
+done
+echo "Clearing maintenance mode on node ${NODE_ID}"
+until [ "${status:-}" = "200" ] || [ "${status:-}" = "400" ]; do
+	status=$(%s)
+	sleep 0.5
+done`, curlNodeIDCommand, curlCommand)
+}
+
+// getPrestopScript creates a script that drains the node before shutting down.
+func (r *SecretResource) getPreStopScript() string {
+	// TODO replace scripts with proper RPK calls
+	curlNodeIDCommand := r.composeCURLGetNodeIDCommand("--silent --fail")
+	curlMaintenanceCommand := r.composeCURLMaintenanceCommand(`-X PUT --silent -o /dev/null -w "%{http_code}"`, nil)
+	genericMaintenancePath := "/v1/maintenance"
+	curlGetMaintenanceCommand := r.composeCURLMaintenanceCommand(`--silent`, &genericMaintenancePath)
+	// TODO(joe): The command builder shouldn't add auth/tls flags, they should be added to an environment variable
+	// in the StatefulSet otherwise a race condition could occur where the mounted script could be updated before
+	// the new setting has been configured on the pod's redpanda node.
+	return fmt.Sprintf(`#!/usr/bin/env bash
+set -e
+
+until NODE_ID=$(%s | grep -o '\"node_id\":[^,}]*' | grep -o '[^: ]*$'); do
+	sleep 0.5
+done
+echo "Setting maintenance mode on node ${NODE_ID}"
+until [ "${status:-}" = "200" ]; do
+	status=$(%s)
+	sleep 0.5
+done
+until [ "${finished:-}" = "true" ] || [ "${draining:-}" = "false" ]; do
```

Comment:
I see we have `terminationGracePeriodSeconds` of 120s. Is that really enough to drain a bigger cluster?